nexus/src/app/saga.rs

```diff
@@ -56,70 +85,91 @@ pub(crate) fn create_saga_dag<N: NexusSaga>(
+    pub(crate) fn new(
+        sec_client: Arc<steno::SecClient>,
+        log: slog::Logger,
+    ) -> SagaExecutor {
+        SagaExecutor { sec_client, log, nexus: OnceLock::new() }
     }
 
-    pub(crate) async fn create_runnable_saga(
-        self: &Arc<Self>,
-        dag: SagaDag,
-    ) -> Result<RunnableSaga, Error> {
-        // Construct the context necessary to execute this saga.
-        let saga_id = SagaId(Uuid::new_v4());
+    // This is a little gross.  We want to hang the SagaExecutor off of Nexus,
+    // but we also need to refer to Nexus, which thus can't exist when
+    // SagaExecutor is constructed.  So we have the caller hand it to us after
+    // initialization.
+    //
+    // This isn't as statically verifiable as we'd normally like.  But it's only
+    // one call site, it does fail cleanly if someone tries to use
+    // `SagaExecutor` before this has been set, and the result is much cleaner
+    // for all the other users of `SagaExecutor`.
+    pub(crate) fn set_nexus(&self, nexus: Arc<Nexus>) {
+        self.nexus.set(nexus).unwrap_or_else(|_| {
+            panic!("concurrent initialization of SagaExecutor")
+        })
+    }
 
-        self.create_runnable_saga_with_id(dag, saga_id).await
+    fn nexus(&self) -> Result<&Arc<Nexus>, Error> {
+        self.nexus
+            .get()
+            .ok_or_else(|| Error::unavail("saga are not available yet"))
     }
 
-    pub(crate) async fn create_runnable_saga_with_id(
-        self: &Arc<Self>,
+    // Low-level interface
+    //
+    // The low-level interface for running sagas starts with `saga_prepare()`
+    // and then uses the `RunnableSaga`, `RunningSaga`, and `StoppedSaga` types
+    // to drive execution forward.
+
+    /// Given a DAG (which has generally been specifically created for a
+    /// particular saga and includes the saga's parameters), prepare to start
+    /// running the saga.  This does not actually start the saga running.
+    ///
+    /// ## Async cancellation
+    ///
+    /// The Future returned by this function is basically not cancellation-safe,
+    /// in that if this Future is cancelled, one of a few things might be true:
+    ///
+    /// * Nothing has happened; it's as though this function was never called.
+    /// * The saga has been created, but not started.  If this happens, the saga
+    ///   will likely start running the next time saga recovery happens (e.g.,
+    ///   the next time Nexus starts up) and then run to completion.
+    ///
+    /// It's not clear what the caller would _want_ if they cancelled this
+    /// future, but whatever it is, clearly it's not guaranteed to be true.
+    /// You're better off avoiding cancellation.  Fortunately, we currently
+    /// execute sagas either from API calls and background tasks, neither of
+    /// which can be cancelled.  **This function should not be used in a
+    /// `tokio::select!` with a `timeout` or the like.**
+    pub(crate) async fn saga_prepare(
```

Comment:
Naming nitpick, for this and `saga_execute` -- this is already defined on `SagaExecutor` struct, so the "saga" part is already implied. 

Most callsites look like:

```rust
        self.sagas.saga_execute::<T>(saga_params).await?;
```

or

```rust
        let runnable_saga = nexus.sagas.saga_prepare(dag).await.unwrap();
```

Wouldn't we still have plenty of context by just calling these `execute` and `prepare`?

Reply:
I read those as "execute" or "prepare" the SagaExecutor itself (since that's the implied subject), which of course isn't right.  I don't feel strongly about it.